source/particle/generator/quadrature_points.cc

```diff
@@ -29,62 +29,85 @@ namespace aspect
-                Point<dim> particle_position_real = (*q_points_itr);
+            n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * fe_values.n_quadrature_points;
+          }
+
+        MPI_Scan(&n_particles_to_generate, &prefix_sum, 1, MPI_UNSIGNED, MPI_SUM, this->get_mpi_communicator());
 
-                try
+        starting_particle_index += prefix_sum - n_particles_to_generate;
+        types::particle_index particle_index = starting_particle_index;
+
+        for (; cell != endc; cell++)
+          {
+            if (cell->is_locally_owned())
+              {
+                fe_values.reinit(cell);
+                std::vector<Point<dim>> quadrature_points = fe_values.get_quadrature_points();
+                for (typename std::vector<Point<dim>>::const_iterator q_points_itr = quadrature_points.begin();
+                     q_points_itr != quadrature_points.end(); q_points_itr++)
                   {
-                    Point<dim> particle_position_unit = this->get_mapping().transform_real_to_unit_cell(cell, (*q_points_itr));
-                    if (GeometryInfo<dim>::is_inside_unit_cell(particle_position_unit))
+                    Point<dim> particle_position_real = (*q_points_itr);
+
+                    try
                       {
-                        const Particle<dim> particle(particle_position_real, particle_position_unit, particle_index);
-                        const types::LevelInd cell_index(cell->level(), cell->index());
-                        particles.insert(std::make_pair(cell_index, particle));
-                        particle_index++;
+                        Point<dim> particle_position_unit = this->get_mapping().transform_real_to_unit_cell(cell, (*q_points_itr));
+                        if (GeometryInfo<dim>::is_inside_unit_cell(particle_position_unit))
+                          {
+                            const Particle<dim> particle(particle_position_real, particle_position_unit, particle_index);
+                            const types::LevelInd cell_index(cell->level(), cell->index());
+                            particles.insert(std::make_pair(cell_index, particle));
+                            particle_index++;
+                          }
+                      }
+                    catch (typename Mapping<dim>::ExcTransformationFailed &)
+                      {
+                        AssertThrow (true,
+                                     ExcMessage ("Couldn't generate particle (unusual cell shape?). "));
                       }
-                  }
-                catch (typename Mapping<dim>::ExcTransformationFailed &)
-                  {
-                    AssertThrow (true,
-                                 ExcMessage ("Couldn't generate particle (unusual cell shape?). "));
                   }
               }
           }
+
+        starting_particle_index = Utilities::MPI::max(particle_index, this->get_mpi_communicator());
```

Comment:
Lastly, update the global starting particle index such that if this function is called again, unique particle indices can still be maintained.